src/FeaturesViewer.cpp

```diff
@@ -588,33 +465,66 @@ namespace qtAliceVision
+          drawFeaturePoint(currentFrameId, previousFrameId, previousFeature, previousPointColor, nbReprojectionErrorLinesDrawn, nbHighlightPointsDrawn, nbPointsDrawn, trackFeatures.nbLandmarks > 0);
 
           // draw track last point
           if (frameId == trackFeatures.maxFrameId)
-            drawFeaturePoint(currentFrameId, frameId, feature, getColor(false, contiguous, currentFeatureInlier, trackHasInliers), nbReprojectionErrorLinesDrawn, nbHighlightPointsDrawn, nbPointsDrawn, trackHasInliers);
+            drawFeaturePoint(currentFrameId, frameId, feature, getPointColor(contiguous, currentFeatureInlier, trackHasInliers), nbReprojectionErrorLinesDrawn, nbHighlightPointsDrawn, nbPointsDrawn, trackFeatures.nbLandmarks > 0);
 
           // draw track line
-          const QColor&  c = getColor(true, contiguous, inliers, trackHasInliers);
-          const unsigned vIdx = nbTrackLinesDrawn * kLineVertices;
+          const QColor&  c = getLineColor(contiguous, inliers, trackHasInliers);
+          unsigned int vIdx = nbTrackLinesDrawn[state] * kLineVertices;
 
+          QPointF prevPoint;
+          QPointF curPoint;
           if (_display3dTracks && trackHasInliers) // 3d track line
           {
-            setVerticeTrackLine(vIdx, QPointF(previousFeature->rx(), previousFeature->ry()), c);
-            setVerticeTrackLine(vIdx + 1, QPointF(feature->rx(), feature->ry()), c);
+            prevPoint = QPointF(previousFeature->rx(), previousFeature->ry());
+            curPoint = QPointF(feature->rx(), feature->ry());
           }
           else // 2d track line
           {
-            setVerticeTrackLine(vIdx, QPointF(previousFeature->x(), previousFeature->y()), c);
-            setVerticeTrackLine(vIdx + 1, QPointF(feature->x(), feature->y()), c);
+            prevPoint = QPointF(previousFeature->x(), previousFeature->y());
+            curPoint = QPointF(feature->x(), feature->y());
           }
 
-          ++nbTrackLinesDrawn;
+          setVertex(verticesTrackLines[state], vIdx, prevPoint, c);
+          setVertex(verticesTrackLines[state], vIdx + 1, curPoint, c);
+
+          ++nbTrackLinesDrawn[state];
 
           previousTrackLineContiguous = contiguous;
+
+          if (_displayTrackEndpoints)
+          {
+            // draw global start point
+            if (previousFrameId == globalTrackInfo.startFrameId)
+            {
+              vIdx = nbEndpointsDrawn * kTriangleVertices;
+              QPointF triangle[] = {QPointF(1, 0), QPointF(-1, 1), QPointF(-1, -1)};
+              float angle = QLineF(curPoint, prevPoint).angle() - rotation();
+              float size = 10.f;
+              auto tr = QTransform().rotate(180 - angle).scale(size, size);
+              for (int i = 0; i < kTriangleVertices; i++) 
+                setVertex(verticesEndpoints, vIdx + i, prevPoint + tr.map(triangle[i]), _endpointColor);
```

Comment:
Track begin/end points can be drawn first in order to avoid missing tracks/points behind them.